# Tidy Data, A Case Study Recreation
This jupyter notebook contains python code to perform the same analysis done in the case study (section 5) of Wickham's Tidy Data article. The original article and code in R can be found here: https://www.jstatsoft.org/article/view/v059i10


The following notebook uses the following packages and their dependencies:<br>
-Numpy<br>
-Pandas<br>

## Base Imports and Packages
We use two main packages to manipulate and tidy the data that we will import:
1. Numpy: Useful for array manipulation, and general math.
2. Pandas: General tools for data analysis, particularly useful to translate R functions into Python.

In [1]:
import numpy as np
import pandas as pd

## Loading data
The data used by Wickham is loaded from its CSV file here. Should any problems arise, the files can be found here: https://www.jstatsoft.org/index.php/jss/article/view/v059i10/2836

The file 'deaths08.csv.bz2' should be extracted and placed besides this file, as well as the file 'icd-main.csv'. This second file contains the codes of the diseases used later.

After loading, we print a full summary of the data. Notice that we have the 539,530 data points mentioned.

In [2]:
mortality = pd.read_csv('deaths08.csv')
codes = pd.read_csv('icd-main.csv')
codes = codes.rename(columns={'code':'cod'})
mortality

yob  mob  dob  sex age_unit  age  nation  marital  stateL  countyL  \
0       1918    0    0    1        A   90       1        2       1        1   
1       1936    0    0    1        A   72       1        2       1        1   
2       1959    0    0    1        A   49       1        5      33        0   
3       1929    0    0    2        A   79       1        5       1        1   
4       1993    0    0    1        A   15       1        1       1        1   
...      ...  ...  ...  ...      ...  ...     ...      ...     ...      ...   
539525  2008   10    5    1        D    1       1        8      32       16   
539526  2008    9   28    2        D   20       1        8      14       25   
539527  2008   11    8    2        D    3       1        8      32       17   
539528  2008    9    1    1        D   24       2        8      32       37   
539529  2008    9   20    1        D    2       1        8      32       57   

        ...  year_reg  mon_reg  day_reg  weight  year_cert  mon_cert  \
0       ...      2008        1        8    8888       2008         1   
1       ...      2008        1       14    8888       2008         1   
2       ...      2008        1       17    8888       2008         1   
3       ...      2008        1       21    8888       2008         1   
4       ...      2008        1        2    8888       2008         1   
...     ...       ...      ...      ...     ...        ...       ...   
539525  ...      2008       10        6    8888       2008        10   
539526  ...      2008       10       18    8888       2008        10   
539527  ...      2008       12       11    8888       2008        12   
539528  ...      2008        9       25    8888       2008         9   
539529  ...      2008        9       22    8888       2008         9   

        day_cert  pregnant  labor_cod  labor_c  
0              7         8          8        8  
1             13         8          8        8  
2             12         8          8        8  
3             20         8          8        8  
4              1         8          8        8  
...          ...       ...        ...      ...  
539525         6         8          8        8  
539526        18         8          8        8  
539527        11         8          8        8  
539528        26         8          8        8  
539529        22         8          8        8  

[539530 rows x 46 columns]

## Inspecting data
We can make use of many of pandas' built-in functions to observe different attributes and qualities of our dataframe.

In [171]:
mortality.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 539530 entries, 0 to 539529
Data columns (total 46 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   yob         539530 non-null  int64 
 1   mob         539530 non-null  int64 
 2   dob         539530 non-null  int64 
 3   sex         539530 non-null  int64 
 4   age_unit    539530 non-null  object
 5   age         539530 non-null  int64 
 6   nation      539530 non-null  int64 
 7   marital     539530 non-null  int64 
 8   stateL      539530 non-null  int64 
 9   countyL     539530 non-null  int64 
 10  locationL   539530 non-null  int64 
 11  popL        539530 non-null  int64 
 12  job         539530 non-null  int64 
 13  edu         539530 non-null  int64 
 14  derhab      539530 non-null  int64 
 15  statD       539530 non-null  int64 
 16  countyD     539530 non-null  int64 
 17  locationD   539530 non-null  int64 
 18  popD        539530 non-null  int64 
 19  placeD      539530 non-

Note: Objects are any non-numeric values within the dataframe. They are often caused by strings being present in the column. This could be a problem if our column is expected to be numeric:

In [172]:
mortality.iloc[456780,39]

8888

What does 'X' mean here? Let's inspect it further!

In [173]:
xmort = mortality[mortality["des"]=="X"]
xmort.describe()

yob           mob           dob           sex           age  \
count  56427.000000  56427.000000  56427.000000  56427.000000  56427.000000   
mean    1926.218034      2.664115      7.422883      1.478672     74.001365   
std      155.228866      3.266881      9.633370      0.500683     66.589774   
min        0.000000      0.000000      0.000000      0.000000      1.000000   
25%     1922.000000      0.000000      0.000000      1.000000     61.000000   
50%     1931.000000      1.000000      0.000000      1.000000     77.000000   
75%     1947.000000      5.000000     15.000000      2.000000     86.000000   
max     2008.000000     12.000000     31.000000      2.000000    998.000000   

             nation       marital        stateL       countyL     locationL  \
count  56427.000000  56427.000000  56427.000000  56427.000000  56427.000000   
mean       1.004094      3.400021     17.111099     59.476492     33.789994   
std        0.063852      2.004040      8.166931     77.455117    141.582004   
min        1.000000      0.000000      1.000000      0.000000      0.000000   
25%        1.000000      2.000000     11.000000     13.000000      1.000000   
50%        1.000000      2.000000     16.000000     36.000000      1.000000   
75%        1.000000      5.000000     24.000000     79.000000     14.000000   
max        2.000000      8.000000     35.000000    570.000000   4733.000000   

       ...  year_reg       mon_reg       day_reg        weight     year_cert  \
count  ...   56427.0  56427.000000  56427.000000  56427.000000  56427.000000   
mean   ...    2008.0      6.466355     15.371081   8880.851330   2007.988002   
std    ...       0.0      3.551762      8.981768    230.461223      0.108876   
min    ...    2008.0      1.000000      0.000000      0.000000   2007.000000   
25%    ...    2008.0      3.000000      8.000000   8888.000000   2008.000000   
50%    ...    2008.0      6.000000     15.000000   8888.000000   2008.000000   
75%    ...    2008.0     10.000000     23.000000   8888.000000   2008.000000   
max    ...    2008.0     12.000000     31.000000   8888.000000   2008.000000   

           mon_cert      day_cert      pregnant     labor_cod       labor_c  
count  56427.000000  56427.000000  56427.000000  56427.000000  56427.000000  
mean       6.495348     15.611055      7.805164      7.613678      7.614723  
std        3.554054      8.874508      0.748277      1.475227      1.470887  
min        0.000000      0.000000      1.000000      1.000000      1.000000  
25%        3.000000      8.000000      8.000000      8.000000      8.000000  
50%        7.000000     16.000000      8.000000      8.000000      8.000000  
75%       10.000000     23.000000      8.000000      8.000000      8.000000  
max       12.000000     31.000000      8.000000      8.000000      8.000000  

[8 rows x 43 columns]

There is some ambiguity whether it means "10" or "NaN", so let us assume it is an error and figure out how to solve it. We can see if there are any strange relationships by scrolling the table and taking a closer look at related variables. Note that "weight" here is 8888, which is clearly indicative of an incorrect value!

In [174]:
non

yob  mob  dob  sex age_unit  age  nation  marital  stateL  countyL  \
3       1929    0    0    2        A   79       1        5       1        1   
7       1905    0    0    1        A  103       1        2       1        1   
16      1948    0    0    2        A   60       1        5       1        2   
17      1998    0    0    2        A   10       1        8       1        2   
22      1920    1   19    1        A   87       1        5       1        2   
...      ...  ...  ...  ...      ...  ...     ...      ...     ...      ...   
539503  2008    4   16    1        D    6       1        8      32       50   
539505  2008    2   11    2        D   15       1        8      32       10   
539506  2008    0    0    2        M    1       1        8      32       17   
539513  2008    0    0    1        M   11       1        8      14       61   
539517  2007    9    0    2        M    4       1        8      32       56   

        ...  year_reg  mon_reg  day_reg  weight  year_cert  mon_cert  \
3       ...      2008        1       21    8888       2008         1   
7       ...      2008        1       10    8888       2008         1   
16      ...      2008        1       14    8888       2008         1   
17      ...      2008        1       28    8888       2008         1   
22      ...      2008        1        3    8888       2007        12   
...     ...       ...      ...      ...     ...        ...       ...   
539503  ...      2008        4       22    8888       2008         4   
539505  ...      2008        2       26    8888       2008         2   
539506  ...      2008        1        5    8888       2008         1   
539513  ...      2008       11       22    8888       2008        11   
539517  ...      2008        1       10    8888       2008         1   

        day_cert  pregnant  labor_cod  labor_c  
3             20         8          8        8  
7              3         8          8        8  
16             9         8          8        8  
17            27         5          2        2  
22            31         8          8        8  
...          ...       ...        ...      ...  
539503        22         8          8        8  
539505        26         8          8        8  
539506         5         8          8        8  
539513        22         8          8        8  
539517        10         8          8        8  

[56427 rows x 46 columns]

## Data Imputation
A common question we may ask is "how to deal with incorrect data", and the answer here is varied. For very few missing data points, it may be possible to just ignore it. For significant amounts of missing data, however, we run the risk of damaging the relationships within our dataset by removing entries. Here, notice we lose about 60,000 entries if we remove these rows -- not a good thing!

Instead, let us see how we can reduce the lost data. One idea is to use the *mean* of the available data. If we have a uniform distribution, this can very accurately replace the missing data.

In [175]:
nonXmort=mortality[mortality["des"]!="X"]
nonXmort

yob  mob  dob  sex age_unit  age  nation  marital  stateL  countyL  \
0       1918    0    0    1        A   90       1        2       1        1   
1       1936    0    0    1        A   72       1        2       1        1   
2       1959    0    0    1        A   49       1        5      33        0   
4       1993    0    0    1        A   15       1        1       1        1   
5       1964    0    0    2        A   44       1        5       1        1   
...      ...  ...  ...  ...      ...  ...     ...      ...     ...      ...   
539525  2008   10    5    1        D    1       1        8      32       16   
539526  2008    9   28    2        D   20       1        8      14       25   
539527  2008   11    8    2        D    3       1        8      32       17   
539528  2008    9    1    1        D   24       2        8      32       37   
539529  2008    9   20    1        D    2       1        8      32       57   

        ...  year_reg  mon_reg  day_reg  weight  year_cert  mon_cert  \
0       ...      2008        1        8    8888       2008         1   
1       ...      2008        1       14    8888       2008         1   
2       ...      2008        1       17    8888       2008         1   
4       ...      2008        1        2    8888       2008         1   
5       ...      2008        1       12    8888       2008         1   
...     ...       ...      ...      ...     ...        ...       ...   
539525  ...      2008       10        6    8888       2008        10   
539526  ...      2008       10       18    8888       2008        10   
539527  ...      2008       12       11    8888       2008        12   
539528  ...      2008        9       25    8888       2008         9   
539529  ...      2008        9       22    8888       2008         9   

        day_cert  pregnant  labor_cod  labor_c  
0              7         8          8        8  
1             13         8          8        8  
2             12         8          8        8  
4              1         8          8        8  
5             11         5          2        2  
...          ...       ...        ...      ...  
539525         6         8          8        8  
539526        18         8          8        8  
539527        11         8          8        8  
539528        26         8          8        8  
539529        22         8          8        8  

[483103 rows x 46 columns]

In [176]:
nonXmort["des"] = pd.to_numeric(nonXmort["des"])
avg_des = nonXmort["des"].mean()
nonXmort["des"].info()
avg_des

<class 'pandas.core.series.Series'>
Index: 483103 entries, 0 to 539529
Series name: des
Non-Null Count   Dtype
--------------   -----
483103 non-null  int64
dtypes: int64(1)
memory usage: 7.4 MB


C:\Users\rafae\AppData\Local\Temp\ipykernel_19924\243796559.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nonXmort["des"] = pd.to_numeric(nonXmort["des"])


5.877144211482851

In [177]:
mortality[mortality["des"]=="X"] = str(avg_des)
mortality["des"] = pd.to_numeric(mortality["des"])
mortality["des"]

C:\Users\rafae\AppData\Local\Temp\ipykernel_19924\450053185.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '5.877144211482851' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  mortality[mortality["des"]=="X"] = str(avg_des)


0         2.000000
1         9.000000
2         0.000000
3         5.877144
4         9.000000
            ...   
539525    0.000000
539526    9.000000
539527    0.000000
539528    0.000000
539529    9.000000
Name: des, Length: 539530, dtype: float64

However, we have an issue: 'des' is a discrete variable, and the mean is continuous. While not necessarily a problem, this is unlike the rest of our data. Another idea is to replace the missing values with the most common values: Using the mode.

In [178]:
del nonXmort, avg_des, xmort

In [179]:
mortality = pd.read_csv('deaths08.csv')

mortality["des"].mode()

0    9
Name: des, dtype: object

By checking the value counts, we can see how skewed our distribution is, as well. This does not bode well: Our distribution is highly skewed towards higher values... and the missing value entry is actually a significant portion of it -- to the point where it could alter our findings!

In [180]:
mortality["des"].value_counts()

des
9    240892
0     68111
X     56427
2     54124
8     24970
1     24657
4     19167
6     18931
3     16650
7     10182
5      5419
Name: count, dtype: int64

In [181]:
mortality[mortality["des"]=="X"] = str(mortality["des"].mode()[0])
mortality["des"].value_counts()

C:\Users\rafae\AppData\Local\Temp\ipykernel_19924\18279265.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '9' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  mortality[mortality["des"]=="X"] = str(mortality["des"].mode()[0])


des
9    297319
0     68111
2     54124
8     24970
1     24657
4     19167
6     18931
3     16650
7     10182
5      5419
Name: count, dtype: int64

In [182]:
mortality = pd.read_csv('deaths08.csv')

In [183]:
mort = mortality[mortality["des"] != "X"]
mort = pd.to_numeric(mort["des"])
mort.median()

8.0

In [184]:
mortality[mortality["des"]=="X"] = str(int(mort.median()))
mortality["des"].value_counts()

C:\Users\rafae\AppData\Local\Temp\ipykernel_19924\3793770050.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '8' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  mortality[mortality["des"]=="X"] = str(int(mort.median()))


des
9    240892
8     81397
0     68111
2     54124
1     24657
4     19167
6     18931
3     16650
7     10182
5      5419
Name: count, dtype: int64

A more accurate way of replacing this value would be to try to mimic the distribution. While we cannot guarantee a well-behaved probabilistic process, we can assume if we had more data it would behave similarly to the data we already have. This means we can randomly generate data points to fit this distribution. 

To do this, we first see what the probability of having any one choice is by dividing the value counts by the total.

In [185]:
mortality = pd.read_csv('deaths08.csv')

In [186]:
mort = mortality[mortality["des"] != "X"]["des"]
mort = pd.to_numeric(mort)
probs = mort.value_counts()/len(mort)
probs = probs.cumsum()
probs

des
9    0.498635
0    0.639621
2    0.751655
8    0.803342
1    0.854381
4    0.894056
6    0.933242
3    0.967707
7    0.988783
5    1.000000
Name: count, dtype: float64

From here, we can take advantage of the np.random library to make a custom function that returns a number associated with a probability. We do this by randomly generating a number between 0 and 1. All the probabilities that are GREATER than this result are kept, and we pick the lowest from there. This basically means that 9 has a 49.9% chance of occurring, 0 has a 63% - 49% = 13% chance of occuring, and so on.

In [187]:
np.random.seed(3981760)
np.random.rand()

0.6057734486091271

In [188]:
np.random.rand()

0.6445979007225147

In [189]:
def dice_roller(input, verbose=False):
    global probs # Use the probability value from another cell
    if input == "X": # We are going to apply this to the whole column, but we only want to replace 'X'
        dice = np.random.rand()  # Generate a random result
        values = probs.index[probs >= dice] # generate the index list of the values with probability greater than the roll
    else:
        return input # If not X, just return the input
    if verbose: # For testing
        print("Dice result: "+str(dice))
        print("=========================")
        print("Probability values:")
        print(probs[probs >= dice])
        print("==========================")
        print("Probability indeces:")
        print(values)
        print("==========================")
        print("Assigned des must then be:")
    return str(values[0]) # Return lowest index of value list

dice_roller("X", verbose=True)   

Dice result: 0.32483102161731303
Probability values:
des
9    0.498635
0    0.639621
2    0.751655
8    0.803342
1    0.854381
4    0.894056
6    0.933242
3    0.967707
7    0.988783
5    1.000000
Name: count, dtype: float64
Probability indeces:
Index([9, 0, 2, 8, 1, 4, 6, 3, 7, 5], dtype='int64', name='des')
Assigned des must then be:


'9'

In [190]:
mortality = pd.read_csv('deaths08.csv')

dataframe.apply() lets us use any arbitrary function on a column

In [191]:
mortality["des"]=mortality["des"].apply(dice_roller)
mortality["des"].value_counts()

des
9    269030
0     76068
2     60506
8     27878
1     27627
4     21353
6     21095
3     18574
7     11334
5      6065
Name: count, dtype: int64

## Some other statistical measures
Although perhaps less useful for data imputation, Standard Deviation, the IQR, Covariance, and Correlation are all useful ways of understanding our data.

Standard deviation is a measure of how disperse our data is. Take a look at the age STD:

In [192]:
mortality["age"].std() # Standard deviation

59.240216228009515

That's a HUGE dispersion! This means that 66% of our data is centered around the mean +/- 59! Let's check our mean:

In [193]:
mortality["age"].mean()

64.49632272533502

This doesn't make a lot of sense: It means 66% of our data is between 5 and 123 years old! This is a classic sign of some very strange outliers that are skewing our data. Let's check what ages are over 99 (which should be extremely rare to non-existant).

In [194]:
mortality.query("age >= 99")

yob  mob  dob  sex age_unit  age  nation  marital  stateL  countyL  \
7       1905    0    0    1        A  103       1        2       1        1   
44      1905    0    0    1        A  103       1        2       1        3   
88      1909    0    0    1        A   99       1        2       1        6   
90         0    0    0    1        A  998       1        0       1        6   
105     1904    0    0    2        A  104       1        2       1        7   
...      ...  ...  ...  ...      ...  ...     ...      ...     ...      ...   
539249  1907    8   10    1        A  101       1        5      32       42   
539256  1906    2   20    1        A  102       1        5      32       44   
539263  1906    6   30    2        A  102       1        2      32       45   
539266  1909    0    0    2        A   99       1        2      32       47   
539299  1909    5   10    2        A   99       1        2      32       49   

        ...  year_reg  mon_reg  day_reg  weight  year_cert  mon_cert  \
7       ...      2008        1       10    8888       2008         1   
44      ...      2008        1       22    8888       2008         1   
88      ...      2008        1       14    8888       2008         1   
90      ...      2008        1       17    8888       2008         1   
105     ...      2008        1       10    8888       2008         1   
...     ...       ...      ...      ...     ...        ...       ...   
539249  ...      2008       12        9    8888       2008        12   
539256  ...      2008       12       24    8888       2008        12   
539263  ...      2008       12       31    8888       2008        12   
539266  ...      2008       12        8    8888       2008        12   
539299  ...      2008       12       16    8888       2008        12   

        day_cert  pregnant  labor_cod  labor_c  
7              3         8          8        8  
44            22         8          8        8  
88            13         8          8        8  
90            16         8          8        8  
105           10         8          8        8  
...          ...       ...        ...      ...  
539249         9         8          8        8  
539256        24         8          8        8  
539263        25         8          8        8  
539266         8         8          8        8  
539299        15         8          8        8  

[6948 rows x 46 columns]

Yikes! That's a lot of possibly incorrect results. For now, let's say we are just going to remove these rows since they're only 7000 out of nearly 550,000. We can use the .query command (or the logical conditions inside the brackets, as before).

In [195]:
mortality = mortality.query("age <= 99")
mortality

yob  mob  dob  sex age_unit  age  nation  marital  stateL  countyL  \
0       1918    0    0    1        A   90       1        2       1        1   
1       1936    0    0    1        A   72       1        2       1        1   
2       1959    0    0    1        A   49       1        5      33        0   
3       1929    0    0    2        A   79       1        5       1        1   
4       1993    0    0    1        A   15       1        1       1        1   
...      ...  ...  ...  ...      ...  ...     ...      ...     ...      ...   
539525  2008   10    5    1        D    1       1        8      32       16   
539526  2008    9   28    2        D   20       1        8      14       25   
539527  2008   11    8    2        D    3       1        8      32       17   
539528  2008    9    1    1        D   24       2        8      32       37   
539529  2008    9   20    1        D    2       1        8      32       57   

        ...  year_reg  mon_reg  day_reg  weight  year_cert  mon_cert  \
0       ...      2008        1        8    8888       2008         1   
1       ...      2008        1       14    8888       2008         1   
2       ...      2008        1       17    8888       2008         1   
3       ...      2008        1       21    8888       2008         1   
4       ...      2008        1        2    8888       2008         1   
...     ...       ...      ...      ...     ...        ...       ...   
539525  ...      2008       10        6    8888       2008        10   
539526  ...      2008       10       18    8888       2008        10   
539527  ...      2008       12       11    8888       2008        12   
539528  ...      2008        9       25    8888       2008         9   
539529  ...      2008        9       22    8888       2008         9   

        day_cert  pregnant  labor_cod  labor_c  
0              7         8          8        8  
1             13         8          8        8  
2             12         8          8        8  
3             20         8          8        8  
4              1         8          8        8  
...          ...       ...        ...      ...  
539525         6         8          8        8  
539526        18         8          8        8  
539527        11         8          8        8  
539528        26         8          8        8  
539529        22         8          8        8  

[533922 rows x 46 columns]

Our mean actually remained pretty close to 64, but check out the STD:

In [196]:
mortality["age"].mean()

61.09943025385731

In [197]:
mortality["age"].std() 

24.650991178369193

That is much more reasonable! Another thing to look at is the Covariance and the Correlation. The Covariance is a measure of how much change is explained by the variation of a variable with respect to another. Higher numbers suggest a strong positive relation, high negative numbers suggest a strong negative relation, and numbers close to 0 suggest no relation.

Correlation, on the other hand, attempts to explain a linear relation between two variables. 1 means a complete positive relation, so when one value increases the other will also always increase. A -1 is the opposite: a complete negative relation. Anywhere in between means some changes can be explained, with the strength being how close it is to 1 or -1. All variables have a correlation of 1 with themselves.

In [198]:
mortality[["yob","age","weight"]].cov()

yob          age         weight
yob     1840.445338  -616.108640   -4069.474454
age     -616.108640   607.671366    3527.602463
weight -4069.474454  3527.602463  662582.688543

In [199]:
mortality[["yob","age","weight"]].corr()

yob       age    weight
yob     1.000000 -0.582588 -0.116535
age    -0.582588  1.000000  0.175803
weight -0.116535  0.175803  1.000000

Finally, let's take a look at IQR, or InterQuartile Range. This value tells us where values are grouped, by finding the 25% smallest and 25% largest parts of the variable. We can either use the quantile() command or the describe() command.

In [200]:
mortality.describe()["age"]

count    533922.000000
mean         61.099430
std          24.650991
min           1.000000
25%          47.000000
50%          67.000000
75%          80.000000
max          99.000000
Name: age, dtype: float64

In [201]:
mortality["age"].quantile([0.25, 0.75])

0.25    47.0
0.75    80.0
Name: age, dtype: float64

By substracting the 25% percentile from the 75% percentile, we are effectively finding the part of our data that composes 50% of our dataset. This is used to find outliers, and it's centered around the median instead of the mean.

In [202]:
80-47

33

In [203]:
80+33*1.5

129.5

## Tidying: count, subset, join
In order to perform the tidying steps, we use pandas' equivalent functions.
1. To count and group by the hour of death and cause of death, we use .groupby(), specifying each column we want. We add the as_index=False to later merge it with the disease descriptions. Finally, we feed that into .size() to summarize by occurrence.
2. We drop any non-occurences using .dropna()
3. We merge hod2 and codes, two dataframes, to add the description of each disease code.
4. We eliminate 'hour 99', which is taken to mean that the time of death was not recorded, by using a filter. This is done with .query('some logic expression').

In [204]:
hod2 = mortality.groupby(['hod', 'cod'], as_index=False)
hod2

In [205]:
hod2 = hod2.size()
hod2

hod  cod  size
0        0  A02     1
1        0  A04     6
2        0  A06     5
3        0  A09    87
4        0  A15     7
...    ...  ...   ...
16671   99  Y33     2
16672   99  Y34    95
16673   99  Y40     1
16674   99  Y57     3
16675   99  Y86     2

[16676 rows x 3 columns]

In [206]:
hod2 = pd.merge(hod2, codes, how='left', on='cod')
hod2


hod  cod  size                                            disease
0        0  A02     1                        Other salmonella infections
1        0  A04     6              Other bacterial intestinal infections
2        0  A06     5                                          Amebiasis
3        0  A09    87  Diarrhea and gastroenteritis of infectious origin
4        0  A15     7                                                NaN
...    ...  ...   ...                                                ...
16747   99  Y33     2        Other specified events, undetermined intent
16748   99  Y34    95             Unspecified event, undetermined intent
16749   99  Y40     1  Adverse effects in therapeutic use, systemic a...
16750   99  Y57     3  Adverse effects in therapeutic use, other and ...
16751   99  Y86     2                        Sequelae of other accidents

[16752 rows x 4 columns]

In [207]:
hod2[hod2["disease"] == 'NaN']

Empty DataFrame
Columns: [hod, cod, size, disease]
Index: []

In [208]:
hod2.iloc[4,3]

nan

In [209]:
hod2.dropna(inplace=True)

In [210]:
hod2 = hod2.query('hod <= 23').reset_index()
hod2

index  hod  cod  size  \
0          0    0  A02     1   
1          1    0  A04     6   
2          2    0  A06     5   
3          3    0  A09    87   
4          5    0  A16    52   
...      ...  ...  ...   ...   
16223  16247   23  Y34    28   
16224  16248   23  Y57     3   
16225  16249   23  Y83     7   
16226  16250   23  Y86    16   
16227  16251   23  Y89     5   

                                                 disease  
0                            Other salmonella infections  
1                  Other bacterial intestinal infections  
2                                              Amebiasis  
3      Diarrhea and gastroenteritis of infectious origin  
4      Respiratory tuberculosis, not confirmed bacter...  
...                                                  ...  
16223             Unspecified event, undetermined intent  
16224  Adverse effects in therapeutic use, other and ...  
16225  Surgical operation and other surgical procedur...  
16226                        Sequelae of other accidents  
16227                  Sequelae of other external causes  

[16228 rows x 5 columns]

## Analysis: Proportion, Frequency
In order to create the proportions and frequencies, we must find an equivalent for ddply and transform.

We can group objects using groupby to create an equivalent of ddply, and perform our transformations there. For example, we generate a dataframe temp to count the occurrence of each hour of death (mortality.groupby('hod').size() ) and we then scale it by its total, then we merge it with hod2. 

To replicate table 16, we need to do the following (reminder: Freq is used as Size in our table as that is the default name that pandas assigns it):
1. Group by hours, obtain the total amount of deaths in that hour, then obtain the proportion. For this, we need to obtain a groupby that counts the frequency of each hour, scale it, and then merge it to our original dataframe.
2. Obtain the overall proportion of the cause of deaths in total and in each hour, add it to the dataframe. We perform the same steps as above, using cod instead of hod. We used an auxiliary column, size_dis, which is computed in the case study on the next step (it is the variable 'n' in the devi dataframe), but we found useful for the generation of the prop column.


In [211]:
### 1 - Group by hours... ###
temp = mortality.groupby('hod', as_index=False).size()
temp = temp.query('hod <= 23') # temp[temp["hod"] <= 23]
temp = temp.dropna()
temp = temp.rename(columns={'size':'size_all'})
temp['prop_all']=temp['size_all']/sum(temp['size_all'])
hod2 = pd.merge(hod2, temp, how='left', on='hod')
hod2

index  hod  cod  size  \
0          0    0  A02     1   
1          1    0  A04     6   
2          2    0  A06     5   
3          3    0  A09    87   
4          5    0  A16    52   
...      ...  ...  ...   ...   
16223  16247   23  Y34    28   
16224  16248   23  Y57     3   
16225  16249   23  Y83     7   
16226  16250   23  Y86    16   
16227  16251   23  Y89     5   

                                                 disease  size_all  prop_all  
0                            Other salmonella infections     20076  0.037990  
1                  Other bacterial intestinal infections     20076  0.037990  
2                                              Amebiasis     20076  0.037990  
3      Diarrhea and gastroenteritis of infectious origin     20076  0.037990  
4      Respiratory tuberculosis, not confirmed bacter...     20076  0.037990  
...                                                  ...       ...       ...  
16223             Unspecified event, undetermined intent     21247  0.040206  
16224  Adverse effects in therapeutic use, other and ...     21247  0.040206  
16225  Surgical operation and other surgical procedur...     21247  0.040206  
16226                        Sequelae of other accidents     21247  0.040206  
16227                  Sequelae of other external causes     21247  0.040206  

[16228 rows x 7 columns]

In [213]:
### 2 - Obtain the overall proportion ###
temp = mortality.groupby('cod', as_index=False).size()
temp.dropna()
temp = temp.rename(columns={'size':'size_dis'})
hod2 = pd.merge(spotify, ymus, how='left', on='nombre_cancion')
hod2['prop'] = hod2['size']/hod2['size_dis']
hod2


KeyError: 'size_dis'

In [221]:
temp = hod2[hod2["hod"]==12]
temp.iloc[400:,:]

index  hod  cod  size  \
8436   8449   12  M79    23   
8437   8450   12  M80     2   
8438   8451   12  M81    11   
8439   8452   12  M86     7   
8440   8453   12  M89     1   
...     ...  ...  ...   ...   
8732   8745   12  Y79     1   
8733   8746   12  Y83     7   
8734   8747   12  Y84     3   
8735   8748   12  Y86    10   
8736   8749   12  Y89     1   

                                                disease  size_all  prop_all  \
8436  Other soft tissue disorders, not elsewhere cla...     23179  0.043862   
8437            Osteoporosis with pathological fracture     23179  0.043862   
8438         Osteoporosis without pathological fracture     23179  0.043862   
8439                                      Osteomyelitis     23179  0.043862   
8440                            Other disorders of bone     23179  0.043862   
...                                                 ...       ...       ...   
8732  Orthopedic devices associated with adverse inc...     23179  0.043862   
8733  Surgical operation and other surgical procedur...     23179  0.043862   
8734  Other medical procedures as the cause of abnor...     23179  0.043862   
8735                        Sequelae of other accidents     23179  0.043862   
8736                  Sequelae of other external causes     23179  0.043862   

      size_dis_x      prop  size_dis_y  
8436         488  0.047131         488  
8437          48  0.041667          48  
8438         128  0.085938         128  
8439          96  0.072917          96  
8440          30  0.033333          30  
...          ...       ...         ...  
8732           4  0.250000           4  
8733         174  0.040230         174  
8734          23  0.130435          23  
8735         365  0.027397         365  
8736          39  0.025641          39  

[301 rows x 10 columns]